server/src/models/timestamps.rs

```diff
@@ -0,0 +1,352 @@
+    // this generally becomes the predominant length of the compressed segments.
+    let mut timestamp_builder = TimestampBuilder::new(compression::GORILLA_MAXIMUM_LENGTH);
+
+    // Add the first timestamp stored as `start_time` in the segment.
+    timestamp_builder.append_value(start_time);
+
+    // Remove the one bit used as a flag to specify that Gorilla is used.
+    let mut bits = BitReader::try_new(residual_timestamps).unwrap();
+    bits.read_bit();
+
+    // Decompress the second timestamp stored as a delta in 14 bits.
+    let mut last_delta = bits.read_bits(14);
+    let mut timestamp = start_time + last_delta as i64;
+    timestamp_builder.append_value(timestamp);
+
+    // Decompress the remaining residual timestamps.
+    while !bits.is_empty() {
+        // Read the next flag with the value of 0, 10, 110, 1110, or 1111.
+        let mut leading_one_bits = 0;
+        while leading_one_bits < 4 && !bits.is_empty() && bits.read_bit() {
+            leading_one_bits += 1;
+        }
+
+        // Any leftover bits in residual_timestamps are set to one. Thus, a
+        // sequence of one bits followed by fewer bits than specified by the
+        // flag means that all residual timestamps have been decompressed.
+        if leading_one_bits != 0 && bits.remaining_bits() < 7 {
+            break;
+        }
+
+        let delta = match leading_one_bits {
+            0 => last_delta,                                               // Flag is 0.
+            1 => read_decode_and_compute_delta(&mut bits, 7, last_delta),  // Flag is 10.
+            2 => read_decode_and_compute_delta(&mut bits, 9, last_delta),  // Flag is 110.
+            3 => read_decode_and_compute_delta(&mut bits, 12, last_delta), // Flag is 1110.
+            4 => last_delta + bits.read_bits(32),                          // Flag is 1111.
+            _ => panic!("Unknown encoding of timestamps."),
+        };
+
+        timestamp += delta as i64;
+        timestamp_builder.append_value(timestamp);
+        last_delta = delta;
+    }
+
+    // Add the last timestamp stored as `end_time` in the segment.
+    timestamp_builder.append_value(end_time);
+    timestamp_builder.finish()
+}
+
+/// Read the next delta-of-delta as `bits_to_read` from `bits`, decode the
+/// delta-of-delta, and add it to `last_delta` to compute the next delta.
+/// [`read_decode_and_compute_delta`] is implemented based on [code published by
+/// Jerome Froelich] under MIT.
+///
+/// [code published by Jerome Froelich]: code published by Jerome Froelich
+fn read_decode_and_compute_delta(bits: &mut BitReader, bits_to_read: u8, last_delta: u32) -> u32 {
+    let encoded_delta_of_delta = bits.read_bits(bits_to_read);
+    let delta_of_delta = if encoded_delta_of_delta > (1 << (bits_to_read - 1)) {
+        encoded_delta_of_delta | (u32::max_value() << bits_to_read)
+    } else {
+        encoded_delta_of_delta
+    };
+    // Wrapping_add() ensure negative values are handled correctly
```

Comment:
Missing punctuation.

Reply:
Fixed. I also fixed the incorrect use of an uppercase W for the method name.